### ISSUE TITLE
Refactor code editors

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot#28607]
**Godot version:**
3.1.x

**Steps to reproduce:**
script_text_editor.cpp
shader_editor_plugin.cpp
text_editor.cpp

What these 3 files have in common is that they implement literally the same thing 3 times. Some examples:

Same set of enums:
https://github.com/godotengine/godot/blob/21e2419e24b9b868fe5da2735784f31366bd1629/editor/plugins/script_text_editor.h#L99-L104
https://github.com/godotengine/godot/blob/21e2419e24b9b868fe5da2735784f31366bd1629/editor/plugins/shader_editor_plugin.h#L72-L77
https://github.com/godotengine/godot/blob/21e2419e24b9b868fe5da2735784f31366bd1629/editor/plugins/text_editor.h#L64-L69

Same edit menu:
https://github.com/godotengine/godot/blob/21e2419e24b9b868fe5da2735784f31366bd1629/editor/plugins/shader_editor_plugin.cpp#L502-L508
https://github.com/godotengine/godot/blob/21e2419e24b9b868fe5da2735784f31366bd1629/editor/plugins/text_editor.cpp#L577-L584
https://github.com/godotengine/godot/blob/21e2419e24b9b868fe5da2735784f31366bd1629/editor/plugins/script_text_editor.cpp#L1581-L1587

Same menu actions:
https://github.com/godotengine/godot/blob/21e2419e24b9b868fe5da2735784f31366bd1629/editor/plugins/shader_editor_plugin.cpp#L245-L250
https://github.com/godotengine/godot/blob/21e2419e24b9b868fe5da2735784f31366bd1629/editor/plugins/text_editor.cpp#L358-L365
https://github.com/godotengine/godot/blob/21e2419e24b9b868fe5da2735784f31366bd1629/editor/plugins/script_text_editor.cpp#L924-L931

There's muuuch more of it. And the worst part is when you want to add some feature to code editor. You basically need to repeat the code 3 times.

IMO the editors should be refactored to cut the repetitions as much as possible. There are differences here and there, but I bet it's possible to rewrite them in a way that doesn't repeat 80% of lines 3 times.

**Comment by Calinou** [comment]
Closing in favor of https://github.com/godotengine/godot/issues/31739, which has more information about the issue at hand.

@Paulb23 did outstanding work to address this, with more to come in the future: https://github.com/godotengine/godot/pull/40973, https://github.com/godotengine/godot/pull/42775, https://github.com/godotengine/godot/pull/45393, https://github.com/godotengine/godot/pull/49238

**Comment by KoBeWi** [comment]
Reopening, as the PR that will close #31739 doesn't resolve this issue. The refactoring work so far only touched CodeEdit and TextEdit, it's not related to editor classes.

**Comment by akien-mga** [comment]
There have been improvements but there's still more work pending here. Won't happen for 4.0 though, and doesn't need to be tied to a specific milestone as it's internal refactoring.

**Comment by Paulb23** [comment]
This will probably be my main focus for now, given `TextEdit / CodeEdit` are in a somewhat decent place for the time being.

Given this is going to be quite a large refactor, it would be ideal if as part of these changes we can hit several big features / proposals at the the same time.

This is the roughly the current class diagram:
![image](https://user-images.githubusercontent.com/6584330/226182457-d7687484-1f30-46d3-acec-da8b3d598cdc.png)

Several pain points are:
- `ShaderTextEditor` is a `CodeTextEditor`, whereas `ScriptTextEditor`  and `TextEditor` has a `CodeTextEdit` this makes it awkward to support both use cases, and everybody forgets about the Shader editor
- Each editor only edits one type of resource without creating lots of exceptions or duplicating code (#72259) 
- Even within resource types there are variations, currently `ScriptTextEditor` has a lot of GDScript specific code.
- `ShaderEditorPlugin` and `ScriptEditorPlugin` share zero code now they do the same thing (#61459)
- There is currently no way to expand this via plugins (godotengine/godot-proposals/issues/2472)

The main feature, that is currently not possible that I'd like to tackle as part of this, is godotengine/godot-proposals/issues/414 allowing split views and having the same file open multiple times.

Secondly, the `ShaderEditor` is located where is is to allow viewing the scene and editing the Shader at the same time. If this become detachable (godotengine/godot-proposals/issues/28) then this assumption no longer holds. So allowing shaders to be opened in the `ScriptEditor` (godotengine/godot-proposals/issues/5071) would be neat.

Bear in mind this might change as I get into the details, I'm also not expecting to need to create all the classes on the following diagram but should serve as a guide / context for what I would like to move towards:
![image](https://user-images.githubusercontent.com/6584330/226182560-96a4874b-ca2a-4ffe-8644-cd64fba6b49e.png)


`ScriptEditorManager` will be the main view with the script list, managing multiple `ScriptEditorWindows`, any Shader / Script specific code can stay in their respective plugin classes. 

`ScriptEditorWindow` will handle the split view. There will be at least one baked into the `ScriptEditorManager`, others could pop out into a separate windows.

Having the `ScriptEditorResource` will allow multiple `ScriptEditors` to sync changes. It also provides an abstraction layer so `TextEditor` can interact with several resource types without creating a specific editor for it. There still may be exceptions however, in these cases a new editor can extend the existing one's without duplicating everything.

**Comment by Gallilus** [comment]
@Paulb23 have you been able to make some progress on this issue?

**Comment by Paulb23** [comment]
Yes, still under heavy WIP, see: https://github.com/Paulb23/godot/tree/script-editor-resource
